rust/lance-index/src/scalar/inverted/builder.rs

```diff
@@ -69,144 +99,139 @@ impl InvertedIndexBuilder {
         dest_store: &dyn IndexStore,
     ) -> Result<()> {
         self.update_index(new_data).await?;
-        self.save(dest_store).await?;
+        self.write(dest_store).await?;
         Ok(())
     }
 
     #[instrument(level = "debug", skip_all)]
-    async fn update_index(&mut self, mut stream: SendableRecordBatchStream) -> Result<()> {
-        const NUM_SHARDS: usize = 32;
-        let token_maps = (0..NUM_SHARDS)
-            .map(|_| RwLock::new(HashMap::new()))
-            .collect_vec();
+    async fn update_index(&mut self, stream: SendableRecordBatchStream) -> Result<()> {
+        let num_shards = *NUM_SHARDS;
+
         // init the token maps
-        {
-            let mut token_maps = token_maps
-                .iter()
-                .map(|map| map.write().unwrap())
-                .collect_vec();
-            for (token, token_id) in std::mem::take(&mut self.tokens.tokens) {
-                let mut hasher = DefaultHasher::new();
-                hasher.write(token.as_bytes());
-                let shard = hasher.finish() as usize % NUM_SHARDS;
-                token_maps[shard].insert(token, token_id);
-            }
+        let mut token_maps = (0..num_shards).map(|_| HashMap::new()).collect_vec();
+        for (token, token_id) in self.tokens.tokens.iter() {
+            let mut hasher = DefaultHasher::new();
+            hasher.write(token.as_bytes());
+            let shard = hasher.finish() as usize % num_shards;
+            token_maps[shard].insert(token.clone(), *token_id);
+        }
+
+        // spawn workers to build the index
+        let buffer_size = get_num_compute_intensive_cpus()
+            .saturating_sub(num_shards)
+            .max(1)
+            .min(num_shards);
```

Comment:
yes, this buffer is used for IO so we want `NUM_CPUS - NUM_SHARDS`, but higher than `NUM_SHARDS` may not help much because it's faster than that indexing workers, so also limit it to `NUM_SHARDS`.
Added comment to explain this